lib/ide/ide/src/view/view_layout.rs

```diff
@@ -0,0 +1,151 @@
+//! This module contains implementation of ViewLayout with a single TextEditor temporarily
+//! occupying half bottom of the screen temporarily as the default layout.
+
+use wasm_bindgen::prelude::*;
+
+use super::text_editor::TextEditor;
+use super::ui_component::UiComponent;
+
+use basegl::system::web::*;
+use basegl::display::world::World;
+
+use nalgebra::zero;
+use nalgebra::Vector2;
+use std::rc::Rc;
+use std::cell::RefCell;
+use js_sys::Function;
+use web_sys::KeyboardEvent;
+use wasm_bindgen::JsCast;
+
+
+
+//TODO: ViewMode is a temporary enumeration, it will be replaced by proper Panel impl.
+// ================
+// === ViewMode ===
+// ================
+
+#[derive(Clone,Copy,Debug)]
+enum ViewMode {
+    Full,
+    Half
+}
+
+
+
+// ======================
+// === ViewLayoutData ===
+// ======================
+
+#[derive(Debug)]
+struct ViewLayoutData {
+    text_editor      : TextEditor,
+    keyboard_closure : Option<Closure<dyn FnMut(KeyboardEvent)>>,
+    view_mode        : ViewMode,
+    dimensions       : Vector2<f32>
+}
+
+impl Drop for ViewLayoutData {
+    fn drop(&mut self) {
+        self.keyboard_closure.as_ref().map(|keyboard_closure| {
+            let body = document().unwrap().body().unwrap();
+            let callback : &Function = keyboard_closure.as_ref().unchecked_ref();
+            body.remove_event_listener_with_callback("keydown", callback).ok();
+        });
+    }
+}
+
+impl ViewLayoutData {
+    fn set_view_mode(&mut self, view_mode:ViewMode) {
+        self.view_mode = view_mode;
+        self.update();
+    }
+
+    fn switch_mode(&mut self) {
+        if let ViewMode::Half = self.view_mode {
+            self.set_view_mode(ViewMode::Full)
+        } else {
+            self.set_view_mode(ViewMode::Half)
+        }
+    }
+
+    fn set_dimensions(&mut self, dimensions:Vector2<f32>) {
+        self.dimensions = dimensions;
+        self.update();
+    }
+
+    fn update(&mut self) {
```

Comment:
`update` is a very generic name for function.
Perhaps it could be something like `recalculate_layout`? 
(unless this is some kind of convention? seeing that `text_editor.update()` is called

Reply:
I am using `recalculate_layout` instead.